### PR TITLE
Add Middleware to populate source in request context

### PR DESCRIPTION
**Pull request by ryanhall07**
This middleware func populates the request context with the value of the
M3-Source header. Additionally it allows providing an optional
Deserializer to convert the header value into a more user friendly type
for the application layer. Both the typed and raw values are accessible
from the context. Most likely only other middleware will use the raw
bytes to forward along the source across api boundaries.

Additionally the Middleware uses the new request scoped logger to add
the source logging field.

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/query/source/source.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// RawFromContext extracts the raw bytes of the source, or false if it doesn't exist.
+// This is used by middleware to propagate the source across API boundaries. Application code should use FromContext.
+func RawFromContext(ctx context.Context) ([]byte, bool) {
+	b, ok := ctx.Value(rawKey).([]byte)
+	return b, ok
+}
+
+// Middleware adds the headers.SourceHeader value to the request context.
+// Installing this middleware function allows application code to access the typed source value using FromContext.
+// Additionally a source log field is added to the request scope logger.
```
Comment (wesleyk) [inline src/query/source/source.go:82]: 🔥

**Review by wesleyk — APPROVED** [review]

File: src/query/source/source.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (robskillington) [inline src/query/source/source.go:1]: nit: 2021

Comment (ryanhall07) [reply src/query/source/source.go:1]: lol i FINALLY installed the correct template in goland, so this should be the last copyright mistake :)

**Review by robskillington — APPROVED** [review]
LGTM

**Pull request merged**